tests/components/tado/test_service.py

```diff
@@ -13,3 +26,107 @@ async def test_has_services(
     await async_init_integration(hass)
 
     assert hass.services.has_service(DOMAIN, SERVICE_ADD_METER_READING)
+
+
+async def test_add_meter_readings(
+    hass: HomeAssistant,
+) -> None:
+    """Test the add meter reading service."""
+
+    await async_init_integration(hass)
+
+    config_entry: MockConfigEntry = hass.config_entries.async_entries(DOMAIN)[0]
+    fixture: str = load_fixture("tado/add_readings_success.json")
+    with patch(
+        "homeassistant.components.tado.TadoConnector.set_meter_reading",
+        return_value=json.loads(fixture),
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_ADD_METER_READING,
+            service_data={
+                ATTR_CONFIG_ENTRY: config_entry.entry_id,
+                ATTR_READING: 1234,
+            },
+            blocking=True,
+        )
+
+
+async def test_add_meter_readings_exception(
+    hass: HomeAssistant,
+) -> None:
+    """Test the add meter reading service."""
+
+    await async_init_integration(hass)
+
+    config_entry: MockConfigEntry = hass.config_entries.async_entries(DOMAIN)[0]
+    with (
+        patch(
+            "homeassistant.components.tado.TadoConnector.set_meter_reading",
+            return_value=None,
+        ),
+        pytest.raises(ServiceValidationError),
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_ADD_METER_READING,
+            service_data={
+                ATTR_CONFIG_ENTRY: config_entry.entry_id,
+                ATTR_READING: 1234,
+            },
+            blocking=True,
+        )
+
+
+async def test_add_meter_readings_invalid(
+    hass: HomeAssistant,
+) -> None:
+    """Test the add meter reading service."""
+
+    await async_init_integration(hass)
+
+    config_entry: MockConfigEntry = hass.config_entries.async_entries(DOMAIN)[0]
+    fixture: str = load_fixture("tado/add_readings_invalid_meter_reading.json")
+    with (
+        patch(
+            "homeassistant.components.tado.TadoConnector.set_meter_reading",
+            return_value=json.loads(fixture),
+        ),
+        pytest.raises(ServiceValidationError),
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_ADD_METER_READING,
+            service_data={
+                ATTR_CONFIG_ENTRY: config_entry.entry_id,
+                ATTR_READING: 1234,
+            },
+            blocking=True,
+        )
+
```

Comment:
Same here. Nothing is effectively tested: add the proper asserts. :)

Reply:
assertion added